consensus/istanbul/backend/announce.go

```diff
@@ -687,3 +692,193 @@ func (sb *Backend) checkPeersAnnounceVersions() {
+	}
+
+	enodeCertificate := &enodeCertificate{
+		EnodeURL: enodeURL,
+		Version:  version,
+	}
+	enodeCertificateBytes, err := rlp.EncodeToBytes(enodeCertificate)
+	if err != nil {
+		return nil, err
+	}
+	msg := &istanbul.Message{
+		Code:    istanbulEnodeCertificateMsg,
+		Address: sb.Address(),
+		Msg:     enodeCertificateBytes,
+	}
+	// Sign the message
+	if err := msg.Sign(sb.Sign); err != nil {
+		return nil, err
+	}
+	logger.Trace("Generated Istanbul Enode Certificate message", "enodeCertificate", enodeCertificate, "address", msg.Address)
+	return msg, nil
+}
+
+// handleEnodeCertificateMsg handles an enode certificate message.
+// At the moment, this message is only supported if it's sent from a proxied
+// validator to its proxy or vice versa.
+func (sb *Backend) handleEnodeCertificateMsg(peer consensus.Peer, payload []byte) error {
+	logger := sb.logger.New("func", "handleEnodeCertificateMsg")
+
+	var msg istanbul.Message
+	// Decode payload into msg
+	err := msg.FromPayload(payload, istanbul.GetSignatureAddress)
+	if err != nil {
+		logger.Error("Error in decoding received Istanbul Enode Certificate message", "err", err, "payload", hex.EncodeToString(payload))
+		return err
+	}
+	logger = logger.New("msg address", msg.Address)
+	logger.Trace("Handling an Istanbul Enode Certificate message")
+
+	var enodeCertificate enodeCertificate
+	if err := rlp.DecodeBytes(msg.Msg, &enodeCertificate); err != nil {
+		logger.Warn("Error in decoding received Istanbul Enode Certificate message content", "err", err, "IstanbulMsg", msg.String())
+		return err
+	}
+
+	parsedNode, err := enode.ParseV4(enodeCertificate.EnodeURL)
+	if err != nil {
+		logger.Warn("Malformed v4 node in received Istanbul Enode Certificate message", "enodeCertificate", enodeCertificate, "err", err)
+		return err
+	}
+
+	// Handle the special case where this node is a proxy and the proxied validator
+	// sent an enode certificate for the proxy to use in handshakes
+	if sb.config.Proxy && sb.proxiedPeer != nil && sb.proxiedPeer.Node().ID() == peer.Node().ID() && msg.Address == sb.config.ProxiedValidatorAddress {
+		// There may be a difference in the URLv4 string because of `discport`,
+		// so instead compare the ID
+		selfNode := sb.p2pserver.Self()
+		if parsedNode.ID() != selfNode.ID() {
+			logger.Warn("Received Istanbul Enode Certificate message with an incorrect enode url", "message enode url", enodeCertificate.EnodeURL, "self enode url", sb.p2pserver.Self().URLv4())
+			return errors.New("Incorrect enode url")
+		}
+		sb.setEnodeCertificateMsg(&msg)
+		return nil
+	}
+	// TODO: remove this check to allow non-proxy peers to send this message
+	// Issue tracked here: https://github.com/celo-org/celo-blockchain/issues/884
+	if sb.proxyNode == nil || sb.proxyNode.peer == nil || sb.proxyNode.peer.Node().ID() != peer.Node().ID() {
+		logger.Warn("Received Istanbul Enode Certificate message from invalid peer")
+		return errUnauthorizedAnnounceMessage
+	}
+
```

Comment:
wdyt of only executing the code below if this node is a proxied validator (e.g. `sb.config.Proxied && sb.proxyNode != nil && sb.proxyNode.peer != nil && sb.proxyNode.peer.ID() == peer.ID()`)?

That way, this function will only process the message if it's proxy (https://github.com/celo-org/celo-blockchain/pull/873/files#diff-49675a1e6ab38a575ed8c1e9a5f31145R804) or if it's a proxied validator (via the condition that I proposed above in this comment).

For all other cases, it would return an `errUnauthorizedAnnounceMessage` error (and the above condition and code block can be removed).

Reply:
That's what the above if statement checks for-- I'll add `!sb.config.Proxied` to the condition as well to make it clearer

Reply:
OK.  In that case, could you add some a comment on what cases that condition covers as well as the "fall through" case.  

The main reason for my initial comment is that it took me over 15 minutes to figure out what cases those conditions covered, so I think the comments will be very valuable.